### PR TITLE
Adam conway 1801 final

### PR DESCRIPTION
**Pull request by adam-conway**
I felt like this went pretty well. Only hangup is I was able to protect the genre create form from the visitor but knew it wasn't the right way to do it. Eventually figured out the right way however.

### CODE REVIEW

File: app/models/movie.rb [hunk 1]
```diff
@@ -1,5 +1,11 @@
 class Movie < ApplicationRecord
   has_many :actor_movies
   has_many :actors, through: :actor_movies
+  has_many :movie_genres
+  has_many :genres, through: :movie_genres
   belongs_to :director
+
+  def self.average_rating
```
Comment (iandouglas) [inline app/models/movie.rb]: you don't want this to be a class method, nor in the movie class, it'd be better to have this as an instance method for genre, so you could do
```ruby
class MovieGenre ...

  def movie_average
     movies.average(:rating)
  end
end
```

Comment (iandouglas) [reply app/models/movie.rb]: otherwise this would pull the average for ALL movies (because it's a class method)

File: spec/features/visitor_sees_genre_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    genre1 = Genre.create!(name: "Horror")
+    genre2 = Genre.create!(name: "Thriller")
+    genre3 = Genre.create!(name: "Suspense")
+
+    visit genres_path
+
+    expect(page).to have_content(genre1.name)
+    expect(page).to have_content(genre2.name)
+    expect(page).to have_content(genre3.name)
+    expect(page).to_not have_selector(:button, 'Create Genre')
```
Comment (iandouglas) [inline spec/features/visitor_sees_genre_index_spec.rb]: nice!

File: spec/features/visitor_sees_genre_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    average = (movie1.rating + movie2.rating) / 2
+
+    visit genres_path
+    click_on genre1.name
+
+    expect(current_path).to eq(genre_path(genre1))
+    expect(page).to have_content(genre1.name)
+    expect(page).to have_content(movie1.title)
+    expect(page).to have_content(movie2.title)
+    expect(page).to have_content("Average rating: #{average}")
```
Comment (iandouglas) [inline spec/features/visitor_sees_genre_index_spec.rb]: a good check on this would be to add a movie with a very high rating or very low rating, in a different genre, and make sure this average is still calculated correctly ... if these are the only movies you add, then the average will come out exactly as you calculate it on line 43.

File: spec/models/movie_spec.rb [hunk 4]
```diff
@@ -4,4 +4,16 @@
...
     it {is_expected.to have_many(:actors).through(:actor_movies)}
   end
+
+  describe 'activerecord' do
+    it 'returns average rating of movies' do
+      director = Director.create!(name: 'Wakowski Sisters')
+      movie1 = Movie.create!(director_id: director.id, title: "Guardians of the Galaxy", description: "Practically perfect in every way", rating: 5)
+      movie2 = Movie.create!(director_id: director.id, title: "Suicide Squad", description: "Practically perfect in every way", rating: 15)
+
+      average = (movie1.rating + movie2.rating) / 2
```
Comment (iandouglas) [inline spec/models/movie_spec.rb]: yes, but we wanted the average of the movies within a genre, so this should have been in your models/genre_spec.rb instead, with fuller testing around movies in multiple genres to ensure the average was calculated properly.

File: spec/features/visitor_sees_genre_index_spec.rb (hunk 3 above)
Comment (iandouglas) [reply spec/features/visitor_sees_genre_index_spec.rb]: great use of string interpolation here, though!

File: app/controllers/genres_controller.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    @genres = Genre.all
+    @genre = Genre.new
+  end
+
+  def show
+    @genre = Genre.find(params[:id])
+  end
+
+  def create
+    genre = Genre.new(genre_params)
```
Comment (iandouglas) [inline app/controllers/genres_controller.rb]: by NOT making this an instance variable, I think your render edge case would have broken because the form needs `@genre` to build the fields, etc..

File: spec/features/admin_sees_create_new_genre_form_on_index_spec.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    admin = User.create!(username: "admin", password: "password", role: 1)
+    allow_any_instance_of(ApplicationController).to receive(:current_user).and_return(admin)
+
+    visit genres_path
+
+    fill_in "genre[name]", with: "Horror"
+    click_on 'Create Genre'
+
+    expect(page).to have_content("Horror")
+  end
```
Comment (iandouglas) [inline spec/features/admin_sees_create_new_genre_form_on_index_spec.rb]: how would you test your edge case code where a genre did not save correctly?

**Comment by adam-conway** [comment]
Got it, my apologies for the confusion! I just got done doing all the
trips-dashboard stuff for the project so my mind immediately went there.

**Comment by adam-conway** [comment]
Yeah that's correct, the edge case would have failed without it being an
instance variable. To test this, I would have made a separate test where an
admin doesn't fill in anything but clicks the create genre button. I would
then expect that the genre db didn't have anything in it and that I was
still on the genre index
